File: R/token_generation_using_client_credential.R
# This R example shows how to generate an OIDC access token (often a JWT) using client credential method.
# Set up configs in config.json.

library(httr2)
library(jsonlite)

read_auth_config <- function(path = "config.json") {
  if (!file.exists(path)) stop("Config file not found: ", path)

  cfg <- jsonlite::fromJSON(path, simplifyVector = TRUE)
  cfg
}

retrieve_new_access_token <- function(config_path = "config.json") {
  cfg <- read_auth_config(config_path)

  req <- request(cfg$token_url) |>
      req_timeout(30) |>
      req_headers(Accept = "application/json",`Content-Type` = "application/x-www-form-urlencoded") |>
      req_auth_basic(cfg$client_id, cfg$client_secret) |>
      req_body_form(
        grant_type = "client_credentials"
      )
  req <- req |> req_body_form(scope = cfg$scope)
  resp <- req |> req_perform()

  token <- resp_body_json(resp)

  access_token <- token$access_token
  access_token
}

access_token <- retrieve_new_access_token(config_path = "config.json")
print(access_token)

File: python/token_generation_using_client_credential.py
import json
import argparse
import requests
import time
from pathlib import Path

CONFIG_PATH = Path("config.json")

def main():
    # read config JSON file
    cfg = read_token_file(CONFIG_PATH)

    example(cfg)

# read the JSON file
def read_token_file(filepath):
    if not CONFIG_PATH.exists():
            raise FileNotFoundError(f"Auth config file not found: {CONFIG_PATH}")

    with CONFIG_PATH.open("r", encoding="utf-8") as f:
        cfg = json.load(f)
    return cfg


def get_token(cfg):
    payload = {'refresh_token': cfg["refresh_token"], 'grant_type': 'client_credentials', 'scope': cfg["scopes"],
        'client_id':cfg["client_id"], 'client_secret':cfg["client_secret"]}

    r = requests.post(cfg["token_url"], data=payload)

    if r.ok:
        data = r.json()
        print("Token refreshed")
        return data["access_token"]

    else:
        print("Unable to refresh access token. ", r.status_code, r.content)



def example(cfg):
    # get the JWT
    token = get_token(cfg)
    print(token)


if __name__ == '__main__':
    main()
